### PR TITLE
[fix] Ensuring cache warmup respects request form data

### LINKED ISSUE #9664: Jinja filter_values macro does not work with cache warmup
The Jinja [`filter_values`](https://github.com/apache/incubator-superset/blob/955a4fe7925c76b2ffdd838c68099952f9595ed7/superset/jinja_context.py#L31) macro does not work when using the `/superset/warm_up_cache` endpoint as the macro fetches the form-data from the request payload which isn't present. Note I suspect this is also true for the`url_param` which relies on the request payload. 

### Expected results

The `/superset/warm_up_cache` endpoint should accurately mimic exploring a slice either via explorer or from a dashboard.

### Actual results

The request payload is void of form-data.

### Environment

- superset version: `master`
- python version: `3.6`
- node.js version: N/A
- npm version: N/A

### Checklist

Make sure these boxes are checked before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### Additional context

cc: @villebro

### PR DESCRIPTION
**Pull request by john-bodley**
### CATEGORY

Choose one

- [x] Bug Fix
- [ ] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

Fixes https://github.com/apache/incubator-superset/issues/9664.

### ADDITIONAL INFORMATION

- [ ] Has associated issue: https://github.com/apache/incubator-superset/issues/9664
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

to: @dpgaspar @etr2460 @villebro

### CODE REVIEW

**Comment by john-bodley** [comment]
@villebro this a draft fix for https://github.com/apache/incubator-superset/issues/9664. I would love your thoughts on how best to address this problem. Note this solution works, however it seems somewhat egregious to be leveraging the Flask `g` global for this context.

**Review by villebro — APPROVED** [review]

**Comment by villebro** [comment]
@john-bodley I agree that it's not optimal to be passing around `form_data` in many different contexts, but in the short term this is probably the best solution. To make this more robust, perhaps we should introduce a util that has this logic in one central place? There we could also add any necessary documentation to explain why this is necessary.

**Comment by villebro** [comment]
@john-bodley I didn't remember that method; You're right, it should probably go in there. I'm assuming `g` context should only be used if `form_data` is missing in both POST and url params?

Mental note: that method is really bloated/complex and doesn't have any unit tests, should probably be broken up at some point.

**Comment by john-bodley** [comment]
@villebro I agree regarding the bloated/complex nature. I might rename that method `get_form_data_and_slice` and have a `get_form_data` method which merely gets the form-data.

**Comment by john-bodley** [comment]
@villebro I've updated the PR to use the somewhat scary `get_form_data` method. Though the method is complex I do prefer having this all housed in one place as it minimizes potential regressions (which is kind of how we ended up here). 

I've added some basic unit tests to `get_from_data` as well as addition unit tests to the Jinja macros.

File: superset/views/utils.py [hunk 1]
```diff
@@ -101,15 +101,18 @@ def get_viz(
 
 def get_form_data(slice_id=None, use_slice_data=False):
     form_data = {}
-    post_data = request.form.get("form_data")
```
Comment (john-bodley) [inline superset/views/utils.py:104]: @villebro I was spinning my wheels for quite some time trying to get `test_request_context()` to work and then realize that this isn't explicitly related to the POST method and not is the body of the request a JSON object, i.e., the `form_data` field is actually further encoded as a JSON object. 

The form-data is messy and we probably should strive to standardize this at some stage. For now I've renamed `post_data` as `request_form_data` which is more accurate.

File: superset/views/utils.py [hunk 2]
```diff
@@ -101,15 +101,18 @@ def get_viz(
 
 def get_form_data(slice_id=None, use_slice_data=False):
```
Comment (villebro) [inline superset/views/utils.py:102]: While we're at it, let's put in type annotations here; I was surprised to see that this was returning a tuple.

**Review by villebro — APPROVED** [review]
image

This really improves code robustness/readability/quality! I'm surprised this worked out so well, I was expecting more trouble to surface as `g` was added to `get_form_data`. If it doesn't completely blow up `mypy`, adding types to `get_form_data` would really put the icing on this PR.

File: superset/views/utils.py [hunk 3]
```diff
@@ -99,23 +99,28 @@ def get_viz(
...
+    if not form_data and hasattr(g, "form_data"):
+        form_data = getattr(g, "form_data")
+
     url_id = request.args.get("r")
     if url_id:
         saved_url = db.session.query(models.Url).filter_by(id=url_id).first()
         if saved_url:
             url_str = parse.unquote_plus(
-                saved_url.url.split("?")[1][10:], encoding="utf-8", errors=None
+                saved_url.url.split("?")[1][10:], encoding="utf-8"
```
Comment (john-bodley) [inline superset/views/utils.py:123]: @villebro `unquote_plus` was raising a typing issue as it expected `errors` to be of type `str`. Per [here](https://github.com/python/cpython/blob/master/Lib/urllib/parse.py#L653) it seems if `errors is None` it gets set to `replace` which is the default, thus there's no need to specify it.

**Pull request merged**